server/remote_cache/scorecard/scorecard.go

```diff
@@ -62,46 +68,156 @@ func GetCacheScoreCard(ctx context.Context, env environment.Env, req *capb.GetCa
 		nextPageToken = next
 	}
 
+	sortResults(results, &sortOpts{
+		GroupByAction: req.GroupByAction,
+		OrderBy:       req.OrderBy,
+		Descending:    req.Descending,
+	})
 	return &capb.GetCacheScoreCardResponse{
-		Results:       scorecard.Results[start:end],
+		Results:       results[start:end],
 		NextPageToken: nextPageToken,
 	}, nil
 }
 
-// SortResults sorts scorecard results, grouping by actions sorted by min
-// request start timestamp, and sorting the requests within each action by start
-// time.
-// TODO(bduffany): More sorting options
-func SortResults(results []*capb.ScoreCard_Result) {
-	// Compute the min result start for each action. (Each result belongs to
-	// an action, identified by action ID).
-	minStartTime := make(map[string]time.Time)
+func filterResults(results []*capb.ScoreCard_Result, req *capb.GetCacheScoreCardRequest) ([]*capb.ScoreCard_Result, error) {
+	mask := req.GetFilter().GetMask()
+	if len(mask.GetPaths()) == 0 {
+		return results, nil
+	}
+	if !mask.IsValid(req.GetFilter()) {
+		return nil, status.InvalidArgumentErrorf("invalid field mask: paths %s", mask.GetPaths())
+	}
+	predicates := make([]func(*capb.ScoreCard_Result) bool, 0, len(mask.GetPaths()))
+	for _, path := range mask.GetPaths() {
+		switch path {
+		case "cache_type":
+			predicates = append(predicates, func(result *capb.ScoreCard_Result) bool {
+				return result.GetCacheType() == req.GetFilter().GetCacheType()
+			})
+		case "request_type":
+			predicates = append(predicates, func(result *capb.ScoreCard_Result) bool {
+				return result.GetRequestType() == req.GetFilter().GetRequestType()
+			})
+		case "response_type":
+			switch req.GetFilter().GetResponseType() {
+			case capb.ResponseType_OK:
+				predicates = append(predicates, func(result *capb.ScoreCard_Result) bool {
+					return result.GetStatus().GetCode() == int32(codes.OK)
+				})
+			case capb.ResponseType_NOT_FOUND:
+				predicates = append(predicates, func(result *capb.ScoreCard_Result) bool {
+					return result.GetStatus().GetCode() == int32(codes.NotFound)
+				})
+			case capb.ResponseType_ERROR:
+				predicates = append(predicates, func(result *capb.ScoreCard_Result) bool {
+					c := result.GetStatus().GetCode()
+					return c != int32(codes.OK) && c != int32(codes.NotFound)
+				})
+			default:
+				return nil, status.InvalidArgumentErrorf("invalid response type %d", req.GetFilter().GetResponseType())
+			}
+		case "search":
+			predicates = append(predicates, func(result *capb.ScoreCard_Result) bool {
+				s := req.GetFilter().GetSearch()
+				return strings.Contains(result.GetActionId(), s) ||
+					strings.Contains(result.GetActionMnemonic(), s) ||
+					strings.Contains(result.GetTargetId(), s) ||
+					strings.Contains(result.GetDigest().GetHash(), s)
+			})
+		default:
+			return nil, status.InvalidArgumentErrorf("invalid field path %q", path)
+		}
+	}
+
+	out := make([]*capb.ScoreCard_Result, 0, len(results))
```

Comment:
nit: i think it would look slightly cleaner if you defined a separate func to check the predicates. something like
matches := func(r *capb.ScoreCard) bool {...}

then the loop below would become
```
for _, result := range results {
  if(matches(result)) {
    out = append(out, result)
  }
}
```

Reply:
Much cleaner -- done